### ISSUE TITLE
vehicle index switch seems to not have any effect

### ISSUE THREAD

**Issue by rmyadsk** [mseminatore/TeslaJS#234]
I have two vehicles under one account and cannot seem to make any of the samples execute against my second vehicle. I think I may be entering the index switch incorrectly?

root@localhost:~/node/node_modules/teslajs/samples# node vehicleConfig.js -i [indexValue]

I first figured this was a zero-based index, so I tried 1, but the script (let's begin with vehicleConfig.js) returned values for the first vehicle. I then tried 2...same result. I then perused products.js and tried the following vehicle-specific values:

id
vehicle_id
id_s

In all cases the script returns data for the first vehicle. Can someone explain the meaning of the index, i.e. how a user should be calling the second vehicle in an account?

Readme.md has 66 mentions of "id" due to its use in specifying a vehicle, but no explanation of what that index value pertains to - what that value should be - so this would be a good opportunity for a brief update.

**Comment by rmyadsk** [comment]
Update: it seems that either the samples do not transmit vehicle index to the sampleFramework, or the sampleFramework chooses to ignore it. Changing line 49:

`            var vehicle = vehicles[program.index || 0];`

to:

`            var vehicle = vehicles[program.index || 1];`

results in data for my second vehicle being returned.

Is this a bug?

**Comment by rmyadsk** [comment]
>PING

Is anyone watching this board?

**Comment by mseminatore** [comment]
Yes, this project is monitored daily.

**Comment by mseminatore** [comment]
Have you looked through the code and tried debugging the issue?

**Comment by rmyadsk** [comment]
Yes - sampleFramework.js appears to read the index as part of `program` on line 28 but by line 49 it's lost in translation - changing the default value from 0 to 1 effects positive change.

**Comment by mseminatore** [comment]
Are you passing the `-i` parameter on the command line? When I do so I see that `program.index` on line 49 contains the value passed in. If you don't specify the parameter then `program.index` will be undefined and the value used for the car index will be the right side of the `||` on line 49.

**Comment by rmyadsk** [comment]
I am, as so: 

> root@localhost:~/node_modules/teslajs/samples# node vehicleConfig.js -i 1

So above line 49 I inserted the following:

`console.log(program.index);`

The program. now echos:

> undefined

Before reporting on the first vehicle.

**Comment by mseminatore** [comment]
I insert the same code just before line 49 and I see 0 for `-i 0` and I see 1 for `-i 1`. I tested on both Windows and Mac. My command line:

`$node soc -i 1`

Are you using the latest version of the code and have you `npm update` to ensure dependencies are up to date?

**Comment by rmyadsk** [comment]
Yes, most recent install from here via npm. Did an npm update for good measure. Ubuntu 20.04. I decided to go back to a sample script and put some debug in. vehicleConfig.js, before sample.run();:

`console.log(program);`

If I'm not mistaken `program` is picking it up:   `_optionValues: { index: 1 },`

___

> root@localhost:~/node_modules/teslajs/samples# node vehicleConfig -i 1
Command {
  _events:
   [Object: null prototype] {
     'option:username': [Function],
     'option:password': [Function],
     'option:index': [Function],
     'option:uri': [Function] },
  _eventsCount: 4,
  _maxListeners: undefined,
  commands: [],
  options:
   [ Option {
       flags: '-u, --username [string]',
       description: 'username (needed only if token not cached)',
       required: false,
       optional: true,
       variadic: false,
       mandatory: false,
       short: '-u',
       long: '--username',
       negate: false,
       defaultValue: undefined,
       defaultValueDescription: undefined,
       parseArg: undefined,
       hidden: false,
       argChoices: undefined },
     Option {
       flags: '-p, --password [string]',
       description: 'password (needed only if token not cached)',
       required: false,
       optional: true,
       variadic: false,
       mandatory: false,
       short: '-p',
       long: '--password',
       negate: false,
       defaultValue: undefined,
       defaultValueDescription: undefined,
       parseArg: undefined,
       hidden: false,
       argChoices: undefined },
     Option {
       flags: '-i, --index <n>',
       description: 'vehicle index (first car by default)',
       required: true,
       optional: false,
       variadic: false,
       mandatory: false,
       short: '-i',
       long: '--index',
       negate: false,
       defaultValue: undefined,
       defaultValueDescription: undefined,
       parseArg: [Function: parseInt],
       hidden: false,
       argChoices: undefined },
     Option {
       flags: '-U, --uri [string]',
       description: 'URI of test server (e.g. http://127.0.0.1:3000)',
       required: false,
       optional: true,
       variadic: false,
       mandatory: false,
       short: '-U',
       long: '--uri',
       negate: false,
       defaultValue: undefined,
       defaultValueDescription: undefined,
       parseArg: undefined,
       hidden: false,
       argChoices: undefined } ],
  parent: null,
  _allowUnknownOption: false,
  _allowExcessArguments: true,
  _args: [],
  rawArgs:
   [ '/usr/bin/node',
     '/root/node_modules/teslajs/samples/vehicleConfig',
     '-i',
     '1' ],
  _scriptPath: '/root/node_modules/teslajs/samples/vehicleConfig',
  _name: 'vehicleConfig',
  _optionValues: { index: 1 },
  _storeOptionsAsProperties: false,
  _actionResults: [],
  _actionHandler: null,
  _executableHandler: false,
  _executableFile: null,
  _defaultCommandName: null,
  _exitCallback: null,
  _aliases: [],
  _combineFlagAndOptionalValue: true,
  _description: '',
  _argsDescription: undefined,
  _enablePositionalOptions: false,
  _passThroughOptions: false,
  _outputConfiguration:
   { writeOut: [Function: writeOut],
     writeErr: [Function: writeErr],
     getOutHelpWidth: [Function: getOutHelpWidth],
     getErrHelpWidth: [Function: getErrHelpWidth],
     outputError: [Function: outputError] },
  _hidden: false,
  _hasHelpOption: true,
  _helpFlags: '-h, --help',
  _helpDescription: 'display help for command',
  _helpShortFlag: '-h',
  _helpLongFlag: '--help',
  _addImplicitHelpCommand: undefined,
  _helpCommandName: 'help',
  _helpCommandnameAndArgs: 'help [command]',
  _helpCommandDescription: 'display help for command',
  _helpConfiguration: {},
  program: [Circular],
  Command: [Function: Command],
  Option: [Function: Option],
  CommanderError: [Function: CommanderError],
  InvalidOptionArgumentError: [Function: InvalidOptionArgumentError],
  Help: [Function: Help],
  args: [] }

**Comment by rmyadsk** [comment]
And here it is again when dumped to the console from inside sampleFramework.js:

> root@bauraulac:~/node_modules/teslajs/samples# node vehicleConfig -i 1
Command {
  _events:
   [Object: null prototype] {
     'option:username': [Function],
     'option:password': [Function],
     'option:index': [Function],
     'option:uri': [Function] },
  _eventsCount: 4,
  _maxListeners: undefined,
  commands: [],
  options:
   [ Option {
       flags: '-u, --username [string]',
       description: 'username (needed only if token not cached)',
       required: false,
       optional: true,
       variadic: false,
       mandatory: false,
       short: '-u',
       long: '--username',
       negate: false,
       defaultValue: undefined,
       defaultValueDescription: undefined,
       parseArg: undefined,
       hidden: false,
       argChoices: undefined },
     Option {
       flags: '-p, --password [string]',
       description: 'password (needed only if token not cached)',
       required: false,
       optional: true,
       variadic: false,
       mandatory: false,
       short: '-p',
       long: '--password',
       negate: false,
       defaultValue: undefined,
       defaultValueDescription: undefined,
       parseArg: undefined,
       hidden: false,
       argChoices: undefined },
     Option {
       flags: '-i, --index <n>',
       description: 'vehicle index (first car by default)',
       required: true,
       optional: false,
       variadic: false,
       mandatory: false,
       short: '-i',
       long: '--index',
       negate: false,
       defaultValue: undefined,
       defaultValueDescription: undefined,
       parseArg: [Function: parseInt],
       hidden: false,
       argChoices: undefined },
     Option {
       flags: '-U, --uri [string]',
       description: 'URI of test server (e.g. http://127.0.0.1:3000)',
       required: false,
       optional: true,
       variadic: false,
       mandatory: false,
       short: '-U',
       long: '--uri',
       negate: false,
       defaultValue: undefined,
       defaultValueDescription: undefined,
       parseArg: undefined,
       hidden: false,
       argChoices: undefined } ],
  parent: null,
  _allowUnknownOption: false,
  _allowExcessArguments: true,
  _args: [],
  rawArgs:
   [ '/usr/bin/node',
     '/root/node_modules/teslajs/samples/vehicleConfig',
     '-i',
     '1' ],
  _scriptPath: '/root/node_modules/teslajs/samples/vehicleConfig',
  _name: 'vehicleConfig',
  _optionValues: { index: 1 },
  _storeOptionsAsProperties: false,
  _actionResults: [],
  _actionHandler: null,
  _executableHandler: false,
  _executableFile: null,
  _defaultCommandName: null,
  _exitCallback: null,
  _aliases: [],
  _combineFlagAndOptionalValue: true,
  _description: '',
  _argsDescription: undefined,
  _enablePositionalOptions: false,
  _passThroughOptions: false,
  _outputConfiguration:
   { writeOut: [Function: writeOut],
     writeErr: [Function: writeErr],
     getOutHelpWidth: [Function: getOutHelpWidth],
     getErrHelpWidth: [Function: getErrHelpWidth],
     outputError: [Function: outputError] },
  _hidden: false,
  _hasHelpOption: true,
  _helpFlags: '-h, --help',
  _helpDescription: 'display help for command',
  _helpShortFlag: '-h',
  _helpLongFlag: '--help',
  _addImplicitHelpCommand: undefined,
  _helpCommandName: 'help',
  _helpCommandnameAndArgs: 'help [command]',
  _helpCommandDescription: 'display help for command',
  _helpConfiguration: {},
  program: [Circular],
  Command: [Function: Command],
  Option: [Function: Option],
  CommanderError: [Function: CommanderError],
  InvalidOptionArgumentError: [Function: InvalidOptionArgumentError],
  Help: [Function: Help],
  args: [] }

**Comment by mseminatore** [comment]
What version of Node are you using? Here are mine:

- Windows: `v12.13.0`
- Mac: `v8.12.0`

If it node the NodeJS version then I don't know what's next. The code works as expected on Windows and Mac. This may be something unique to your environment. Perhaps Ubuntu.

**Comment by rmyadsk** [comment]
v10.19.0 is the latest that's available via default apt channels. But, doing some reading it appears that's a bit laggy. [NodeSource](https://nodesource.com/) seems like a reputable resource - so I've upgraded to v15.12.0.

Unfortunately the behavior persists. sampleFramework refuses to believe program.index is defined. However, pre-emptively setting it with `program.index. 1;` results in success. May be something specific to my system. Closing.
